active-rfcs/0000-build-performance.md

```diff
@@ -0,0 +1,187 @@
+- Astro's build process relies on scanning the rendered HTML and then *update* the HTML as well, to replace assets with the hashed paths in the build.
+- Because of the above, performance has actually regressed in __0.21__, even though it was never the best even before.
+- In order to support SSR in the future we have to move away from page-scanning as the way to find and build assets, since SSR apps by their nature *cannot* be rendered ahead of time.
+
+# Detailed design
+
+## Enforced static use of client directives
+
+The client directives such as `client:load`, `client:idle`, etc will need to be defined in the hydrated component where they are used, and not rendered dynamically. For example the following is __not allowed__:
+
+```astro
+---
+import Clock from '../components/Clock.jsx';
+const { shouldHydrate } = Astro.props;
+const attrs = {
+  'client:idle': shouldHydrate
+};
+---
+<Clock {...attrs} />
+```
+
+We need to know that the site depends on the `client:idle` directive, so that we can *build* the client-side JavaScript needed.
+
+The implementation will be:
+
+1. In the compiler, include the used directive as part of the exported metadata about the component.
+2. In the compiler, mark the component as having statically included a directive.
+  - How to mark is up to the implementer, but we have other metadata attached to hydrated component usage already, and it would make sense to follow this same method.
+3. When rendering, if a component contains a client directive, make sure the directive is matched by the marking in __(2)__.
+4. If the marking is not there, we know that the directive must have been added statically. Throw an `Error` message for this, letting the user know that the directive must be added statically.
+
+## The `local:` directive
+
+A new `local:` directive will be introduced that can be used to set relative links on any element. The syntax looks like this:
+
+```astro
+<img local:src="../images/penguin.png" />
+```
+
+The implementation is:
+
+1. The compiler will see the `local:` directives. They must be used *statically* and not added dynamically (they will be ignored if added dynamically).
+1. The compiler will convert the value of the directive into an import statement. For example the above becomes:
+
+    ```astro
+    ---
+    import imgUrl from "../images/penguin.png";
+    ---
+    <img local:src={imgUrl} />
+    ```
+1. If the value of the directive is *not* a static string, the compiler should ignore that usage and not add an import statement.
+1. The compiler include a warning about the non-static usage of the `local:` directive.
+
+There are a couple of special cases that need to be accounted for:
+
+- `<img srcset>` value is more complex than just a single URL. We need to parse the value and add multiple import statements.
```

Comment:
can we leave this as post-v1.0? Just making sure that "to be accounted for" doesn't mean that is being marked as "will implement" by this RFC and before v1.0.

Reply:
For what reason? My thinking is that `local:` should work on any element and might be unintuitive if it doesn't. For the most part the compiler doesn't need to know/care what type of element it is. This is one exception, however.

Reply:
`srcset` is more complex than `src` right? I guess more details about how this would work would address this concern, if you think it's reasonable to tackle in this RFC. I'm just unclear how our compiler would parse and handle something like:

```
<img local:srcset="./elva-fairy-480w.jpg 480w,
             ./elva-fairy-800w.jpg 800w"
     sizes="(max-width: 600px) 480px,
            800px"
     local:src="./elva-fairy-800w.jpg"
     alt="Elva dressed as a fairy">
```